File: Python/_459_Repeated_Substring_Pattern/solution.py
class Solution:
    def repeatedSubstringPattern(self, s):
        """
        :type s: str
        :rtype: bool
        """
        i = 1
        j = 0
        dp = [0] * (len(s) + 1)
        while i < len(s): 
            if s[i] == s[j]:
                i += 1
                j += 1
                dp[i] = j
            elif j == 0:
                i += 1
            else:
                j = dp[j]
        return dp[len(s)] > 0 and (dp[len(s)] % (len(s) - dp[len(s)]) == 0)

File: Java/_459_Repeated_Substring_Pattern/Solution.java
class Solution {
    public boolean repeatedSubstringPattern(String s) {
        int i = 1;
        int j = 0;
        int[] dp = new int[s.length() + 1];
        while (i < s.length()) {
            if (s.charAt(i) == s.charAt(j)) {
                dp[++i] = ++j;
            }
            else if (j == 0)
                ++i;
            else
                j = dp[j];
        }
        return dp[s.length()] > 0 && (dp[s.length()] % (s.length() - dp[s.length()]) == 0);
    }
}